src/mono/mono/mini/interp/transform.c

```diff
@@ -673,24 +673,59 @@ handle_branch (TransformData *td, int long_op, int offset)
 	if (offset > 0)
 		init_bb_stack_state (td, target_bb);
 
+	if (td->cbb->no_inlining && long_op != MINT_CALL_HANDLER)
+		target_bb->jump_targets--;
```

Comment:
Just wondering, why not just always eliminate dead blocks early if possible?

Presumably even something like this is worth simplifying, particularly for larger block bodies:
```csharp
if (someExpressionThatIsTriviallyConstantFalse)
{
    // logic that can never be hit
}
else
{
    // logic that will always be hit
}
```

I'd expect it reduces memory overhead, improves overall throughput due to less nodes to traverse, and may even make other optimizations easier to identify.

Reply:
👍, I'm not sure exactly what RyuJIT does for this scenario, but @EgorBo would likely know.

The general consideration comes in for code like this: https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs,2007

Where, even completely ignoring inlining, there exists a large block of code that is always dead for Mono today and will always be dead for most non-xarch platforms.

This type of pattern is pretty common outside the BCL in perf critical code, and that is often some of the more complex/infrastructural code for frameworks/libraries.

Reply:
Actually, what is worse in the code sample you provided is the existance of the `do/while` loop. So in addition to traversing all that code, we will still inline.